history_and_modals.md

```diff
@@ -1,158 +1,357 @@
+
+Read on for more details and realistic usage examples.
+
+### Asking for confirmation
+
+There's a more advanced part of this API, which allows asking the user for confirmation if they really want to close the modal. This is useful in cases where, e.g., the modal contains a form with unsaved data. The usage looks like the following:
+
+```js
+watcher.onbeforeclose = e => {
+  if (hasUnsavedData) {
+    e.preventDefault();
   }
-}
-dialog.showModal();
+};
+```
+
+If the event is canceled, then the browser will show some UI saying something like "Close modal? Changes you made may not be saved." For [abuse prevention](#abuse-analysis), this text is _not_ configurable.
+
+Furthermore, unlike `Window`'s `beforeunload` event, this browser UI does _not_ block the event loop. Instead, it simply has the effect of delaying the `close` event until the user says "Yes".
+
+If the user says "No" (i.e., keep the modal open), then the `ModalCloseWatcher` remains active, and can receive future `beforeclose` (and `close`) events. But once the user says "Yes" (i.e., confirms closing the modal), the `ModalCloseWatcher` receives the final `close` event, and becomes inactive.
+
+The exact UI for this confirmation is up to the user agent. It could be a `beforeunload`-style modal dialog, but the resulting "modal on top of a modal" might be ugly. _TODO maybe we can get some mock-ups of better ideas._
+
+### Platform implementation notes
+
+#### Android: interaction with fragment navigation
+
+Consider a modal which contains a lot of text, such as a terms of service with a table of contents, or a preview overlay that shows this very document. Within such a modal, it'd be possible to perform fragment navigation.
+
+For implementations like Android which use the back button as their close signal, the desired user experience is that the back button navigates back through the normal fragment navigation stack first. Only when the user returns to the original state where the modal was opened, will the back button be a close signal.
+
+This does not impact the specification of this feature, since the exact choice of what counts as a close signal is implemenation-defined. But it does impact the implementation, for Android platforms. There, the existence of an active `ModalCloseWatcher` does not suffice to mean that back button presses should be routed to that `ModalCloseWatcher`. Instead, the implementation needs to make sure that the current session history entry is the same as the one that was current when the `ModalCloseWatcher` was constructed. Only when they match should the back button be counted as a close signal.
+
+Note that no such remembrance of the history state applies on desktop, where the back button is not involved. The <kbd>Esc</kbd> key can always send a close signal, even after some fragment navigations.
```

Comment:
This is a little subtle - is the application then responsible on desktop for unwinding the history stack on desktop, vs on mobile where the browser would unwind the history stack until it closes the modal?

Reply:
Oh, good catch.

On mobile, it's not quite that the browser unwinds the history stack; it's more that the user does so, if they use the back button enough times.

On both desktop and mobile, if the modal closes before enough back button presses (e.g., on desktop using <kbd>Esc</kbd>, or on mobile by using the modal's close button) then the history stack is left as-is, potentially with a few fragment navigations on it. Which, in this proposal, would require the application to clean them up.

That's not great. It at least deserves it's own section explaining the subtleties, but I wonder if there's anything better we can do.

An alternate version is that we could declare history-changes-within-modals out of scope/unsupported. In that version, on desktop <kbd>Esc</kbd> always triggers a modal close watcher (like in this version), and on mobile, the back button always triggers a modal close watcher (without tying it to the current history entry like this version does). Then the application author has to clean up the history state on both desktop and mobile.